sdk/mobile-center-updates/src/main/java/com/microsoft/azure/mobile/updates/Updates.java

```diff
@@ -0,0 +1,971 @@
+            DownloadManager downloadManager = (DownloadManager) mContext.getSystemService(DOWNLOAD_SERVICE);
+            Uri uriForDownloadedFile = downloadManager.getUriForDownloadedFile(mDownloadId);
+            if (uriForDownloadedFile != null) {
+
+                /* Build install intent. */
+                MobileCenterLog.debug(LOG_TAG, "Download was successful for id=" + mDownloadId + " uri=" + uriForDownloadedFile);
+                Intent intent = getInstallIntent(uriForDownloadedFile);
+                boolean installerFound = false;
+                if (intent.resolveActivity(mContext.getPackageManager()) == null) {
+                    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
+                        Cursor cursor = downloadManager.query(new DownloadManager.Query().setFilterById(mDownloadId));
+                        if (cursor != null) {
+                            if (cursor.moveToNext()) {
+                                uriForDownloadedFile = getFileUriOnOldDevices(cursor);
+                                intent = getInstallIntent(uriForDownloadedFile);
+                                installerFound = intent.resolveActivity(mContext.getPackageManager()) != null;
+                            }
+                            cursor.close();
+                        }
+                    }
+                } else {
+                    installerFound = true;
+                }
+                if (!installerFound) {
+                    MobileCenterLog.error(LOG_TAG, "Installer not found");
+                    completeWorkflow(this);
+                    return null;
+                }
+
+                /* Exit check point. */
+                Activity activity;
+                try {
+                    activity = getForegroundActivityWithStateCheck(this);
+                } catch (IllegalStateException e) {
+
+                    /* If we were canceled, exit now. */
+                    return null;
+                }
+
+                /* If foreground, execute now, otherwise post notification. */
+                String uri = uriForDownloadedFile.toString();
+                if (activity != null) {
+
+                    /* This start call triggers strict mode violation in U.I. thread so it needs to be done here, and we can't synchronize anymore... */
+                    MobileCenterLog.debug(LOG_TAG, "Application is in foreground, launch install UI now.");
+                    activity.startActivity(intent);
+                    completeWorkflow(this);
+                } else {
+
+                    /* Remember we have a download ready. */
+                    MobileCenterLog.debug(LOG_TAG, "Application is in background, post a notification.");
+
+                    /* And notify. */
+                    int icon;
+                    try {
+                        ApplicationInfo applicationInfo = mContext.getPackageManager().getApplicationInfo(mContext.getPackageName(), 0);
```

Comment:
It's mandatory in notification manager to have an icon, you can't post the notification otherwise. In reality it will never fail as I use the launcher icon.